### PR TITLE
Refactor: RuleList

### PR DESCRIPTION
**Pull request by kaicataldo**
I split the tests up into their own commit because this got kind of big.

All the tests are passing in the first commit - I just wanted to add more unit tests for extra confidence/documentation for others.

The big thing this achieves is that `Engine` is now responsible for processing rules and managing the stylesheet, and `RuleList` is responsible for the logic of inserting rules.

Removing the auto-processing on insert in my previous PR means that inserting rules no longer has to add placeholder rules to the stylesheet - instead, we insert the rules into the stylesheet on `process` after we have already generated all the rules.

**TL;DR**: `insert`ing rules only adds them to the internally managed rule list, while calling `process` or `toString` takes data, generates all the rules, and manages the removal/insertion into the stylesheet or returns a string of the generated styles, respectively.

### CODE REVIEW

**Comment by weikinhuang** [comment]
:+1:

File: src/Engine.js [hunk 1]
```diff
@@ -1,68 +1,69 @@
...
+  const otherVendorPrefixesMap = {
     moz: ['webkit', 'ms'],
     webkit: ['moz', 'ms'],
     ms: ['webkit', 'moz'],
   };
-  var ua = navigator.userAgent.toLowerCase();
-  var vendorPrefix = ua.indexOf('trident') > -1 ? 'ms' : (ua.indexOf('webkit') > -1 ? 'webkit' : 'moz');
-  var otherVendorPrefixes = otherVendorPrefixesMap[vendorPrefix];
+  const ua = navigator.userAgent.toLowerCase();
+  const vendorPrefix = ua.indexOf('trident') > -1 ? 'ms' : (ua.indexOf('webkit') > -1 ? 'webkit' : 'moz');
```
Comment (designjockey) [inline src/Engine.js]: Interesting, what's `trident`?

Comment (kaicataldo) [inline src/Engine.js]: Internet Explorer!

File: src/Engine.js [hunk 2]
```diff
@@ -1,68 +1,69 @@
...
 
   destroy() {
-    this.rules.length = 0;
-
     if (this.style) {
       this.style.destroy();
     }
   },
 
+  insert(selector, spec) {
```
Comment (designjockey) [inline src/Engine.js]: consider renaming to `insertRule`

File: src/Engine.js (hunk 1 above)
Comment (designjockey) [inline src/Engine.js]: Lol of course.

File: src/Engine.js (hunk 2 above)
Comment (kaicataldo) [inline src/Engine.js]: My preference is not to do this in this PR, but if you feel strongly about it I'm not against it and can change that in a follow up PR. I'm assuming the public API shouldn't change, though, as the public `insert` method can take one rule, an object of rules, and will soon take an array of rules

File: src/RuleList.js [hunk 3]
```diff
@@ -0,0 +1,185 @@
...
+      selector,
+      spec,
+      toggleKeys
+    });
+  }
+
+  _getToggleSelectorInfo(selector) {
+    const toggleKeys = [];
+
+    [toggleSelectorPsuedoRegex, toggleSelectorClassRegex].forEach((toggleSelectorRegex) => {
```
Comment (designjockey) [inline src/RuleList.js]: Type in `toggleSelector**Psuedo**Regex` should be `toggleSelector**Pseudo**Regex`

Comment (kaicataldo) [inline src/RuleList.js]: Damn, nice catch! That's been there for a LONG time haha

File: src/Engine.js (hunk 2 above)
Comment (kaicataldo) [inline src/Engine.js]: Reasoning for not wanting to do that in this PR is that I'm going to have change engine tests, and don't want to obscure the purpose of this PR even more (given the ES5 -> ES6 changes already happening)

File: src/RuleList.js [hunk 4]
```diff
@@ -0,0 +1,185 @@
...
+    return arrayDataFromState.forEach((item, index) => {
+      let { selector, spec, toggleKeys } = descriptor;
+
+      if (filterFunction && !filterFunction(item)) {
+        return;
+      }
+
+      let toggleSuffix = '';
+      selector = selector.replace(arrayPropertyRegex, (match, column) => {
+        toggleSuffix += '_' + column + '_' + item[column];
```
Comment (designjockey) [inline src/RuleList.js]: es6 string format?

File: src/Engine.js (hunk 2 above)
Comment (designjockey) [inline src/Engine.js]: Fair enought

File: test/unit/index.js [hunk 5]
```diff
@@ -14,7 +14,7 @@ describe('Focss', function() {
     }
   });
 
-  it('is instanciable', function() {
+  it('is instantiable', function() {
```
Comment (designjockey) [inline test/unit/index.js]: Lol nice.

File: test/unit/src/RuleList.js [hunk 6]
```diff
@@ -0,0 +1,286 @@
...
+      });
+      this._rules.insert('.bar', {
+        width: 'qux'
+      });
+
+      expect(this._rules.getTraces()).toEqual({});
+    });
+  });
+
+  describe('generateArrayRules', function() {
```
Comment (designjockey) [inline test/unit/src/RuleList.js]: `#generateArrayRules`

**Comment by designjockey** [comment]
👍
